beekeeper-metadata-cleanup/src/test/java/com/expediagroup/beekeeper/metadata/cleanup/service/PagingMetadataCleanupServiceTest.java

```diff
@@ -0,0 +1,285 @@
+
+    tables.forEach(path -> metadataRepository.save(path));
+    pagingCleanupService.cleanUp(Instant.now());
+
+    verify(metadataCleaner, times(2)).dropTable(metadataCaptor.capture());
+    assertThat(metadataCaptor.getAllValues())
+        .extracting("tableName")
+        .containsExactly(tables.get(0).getTableName(), tables.get(1).getTableName());
+    verify(pathCleaner, times(2)).cleanupPath(pathCaptor.capture());
+    assertThat(pathCaptor.getAllValues())
+        .extracting("path")
+        .containsExactly(tables.get(0).getPath(), tables.get(1).getPath());
+  }
+
+  @Test
+  void metadataCleanerException() {
+    Mockito
+        .doNothing()
+        .doThrow(new BeekeeperException("Error"))
+        .when(metadataCleaner)
+        .dropTable(Mockito.any(HousekeepingMetadata.class));
+
+    List<HousekeepingMetadata> tables = List
+        .of(createHousekeepingMetadata("table1", "s3://some_foo", null, SCHEDULED),
+            createHousekeepingMetadata("table2", "s3://some_bar", null, SCHEDULED));
+    tables.forEach(table -> metadataRepository.save(table));
+
+
+    pagingCleanupService.cleanUp(Instant.now());
+
+    verify(metadataCleaner, times(2)).dropTable(metadataCaptor.capture());
+    assertThat(metadataCaptor.getAllValues())
+        .extracting("tableName")
+        .containsExactly(tables.get(0).getTableName(), tables.get(1).getTableName());
+
+    List<HousekeepingMetadata> result = metadataRepository.findAll();
+    assertThat(result.size()).isEqualTo(2);
+    HousekeepingMetadata housekeepingMetadata1 = result.get(0);
+    HousekeepingMetadata housekeepingMetadata2 = result.get(1);
+
+    assertThat(housekeepingMetadata1.getTableName()).isEqualTo(tables.get(0).getTableName());
+    assertThat(housekeepingMetadata1.getPath()).isEqualTo(tables.get(0).getPath());
+    assertThat(housekeepingMetadata1.getHousekeepingStatus()).isEqualTo(DELETED);
+    assertThat(housekeepingMetadata1.getCleanupAttempts()).isEqualTo(1);
+    assertThat(housekeepingMetadata2.getTableName()).isEqualTo(tables.get(1).getTableName());
+    assertThat(housekeepingMetadata2.getPath()).isEqualTo(tables.get(1).getPath());
+    assertThat(housekeepingMetadata2.getHousekeepingStatus()).isEqualTo(FAILED);
+    assertThat(housekeepingMetadata2.getCleanupAttempts()).isEqualTo(1);
+  }
+
+  @Test
+  @Timeout(value = 10)
+  void doNotInfiniteLoopOnRepeatedFailures() {
+    List<HousekeepingMetadata> tables = List
+        .of(createHousekeepingMetadata("table1", "s3://some_foo", null, FAILED),
+            createHousekeepingMetadata("table2", "s3://some_bar", null, FAILED),
+            createHousekeepingMetadata("table3", "s3://some_foobar", null, FAILED));
+
+    for (int i = 0; i < 5; i++) {
+      int finalI = i;
+      tables.forEach(path -> {
+        if (finalI == 0) {
+          metadataRepository.save(path);
+        }
+
+        doThrow(new RuntimeException("Error")).when(metadataCleaner).dropTable(Mockito.any());
```

Comment:
you don't need to mock this all the time I think in the loop, you can just set it outside the for loop